### PR TITLE
Add a method to asynchronously update the expiration cache

### PR DESCRIPTION
**Pull request by 54uso**
Insert "asyncUpdate" into annocation called Cacheable. If "asyncUpdate" is true, when this cache expired, updates data after returning. If 'asyncUpdate' is false, when this cache expired, updates data then returns. The default value of 'asyncUpdate' is false.

### CODE REVIEW

**Comment by dmarkov** [comment]
@54uso Thanks, someone will review your pull request soon

**Comment by dmarkov** [comment]
@pinaf please review

**Comment by pinaf** [comment]
@54uso Is there an issue this PR fixes? If so, please refer to it in your first comment.

**Comment by pinaf** [comment]
@54uso sorry but the design here just doesn't look good. the class is now huge (`GodClass`) - please find an alternative design for the desired functionality. @yegor256 please advise if you disagree with my opinion.

**Comment by yegor256** [comment]
@pinaf I think I agree with you, but the idea implemented here looks good... maybe we can improve the design a bit (I'm not sure we can totally refactor it)?

**Comment by pinaf** [comment]
@54uso do you think you can extract the new functionality into a decorator class?

**Comment by 54uso** [comment]
@pinaf  OK, I'll try my best.

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 1]
```diff
@@ -72,34 +70,19 @@
      * @checkstyle LineLength (2 lines)
      */
     private final transient ConcurrentMap<MethodCacher.Key, MethodCacher.Tunnel> tunnels =
-        new ConcurrentHashMap<MethodCacher.Key, MethodCacher.Tunnel>(0);
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso let's revert this change, keeping the full variable path.

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 2]
```diff
@@ -72,34 +70,19 @@
...
+        new ConcurrentHashMap<Key, Tunnel>(0);
 
     /**
-     * Service that cleans cache.
+     * Save the keys which need update.
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso how about just `Keys that need updating.`?

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 3]
```diff
@@ -170,6 +168,7 @@ public Object flush(final ProceedingJoinPoint point) throws Throwable {
      *
      * @param point Joint point
      * @since 0.7.14
+     * @checkstyle MethodsOrderCheck (3 lines)
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso why this?

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 4]
```diff
@@ -189,6 +188,7 @@ public void preflush(final JoinPoint point) {
      *
      * @param point Joint point
      * @since 0.7.18
+     * @checkstyle MethodsOrderCheck (3 lines)
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso why this?

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 5]
```diff
@@ -148,13 +136,23 @@ public Object cache(final ProceedingJoinPoint point) throws Throwable {
     }
 
     /**
+     * Whether create a new Tunnel.
+     * @param tunnel MethodCacher.Tunnel
+     * @return Boolean
+     */
+    private boolean isCreateTunnel(final MethodCacher.Tunnel tunnel) {
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso private methods should be the last ones in the class.

File: src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        this.updater.schedule(
+            new VerboseRunnable(
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        UpdateMethodCacher.this.update();
+                    }
+                }
+            ),
+            0L, TimeUnit.SECONDS
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java]: @54uso 0 seconds?

File: src/test/java/com/jcabi/aspects/CacheableTest.java [hunk 7]
```diff
@@ -74,6 +74,32 @@ public void cachesSimpleCall() throws Exception {
...
+        final CacheableTest.Foo foo = new CacheableTest.Foo(1L);
+        final String first = foo.asyncGet().toString();
+        final String second = foo.asyncGet().toString();
+        MatcherAssert.assertThat(first, Matchers.equalTo(second));
+        try {
+            TimeUnit.SECONDS.sleep(2);
+        } catch (final InterruptedException ex) {
+            throw new IllegalStateException(ex);
+        }
+        final String thrid = foo.asyncGet().toString();
```
Comment (pinaf) [inline src/test/java/com/jcabi/aspects/CacheableTest.java]: @54uso `third`

File: src/test/java/com/jcabi/aspects/CacheableTest.java [hunk 8]
```diff
@@ -74,6 +74,32 @@ public void cachesSimpleCall() throws Exception {
...
 
     /**
+     * Cacheable can Asynchronous update.
+     * @throws Exception If something goes wrong
+     */
+    @Test
+    public void asyncUpdateCacheSimpleCall() throws Exception {
+        final CacheableTest.Foo foo = new CacheableTest.Foo(1L);
+        final String first = foo.asyncGet().toString();
+        final String second = foo.asyncGet().toString();
```
Comment (pinaf) [inline src/test/java/com/jcabi/aspects/CacheableTest.java]: @54uso we can inline this variable

File: src/test/java/com/jcabi/aspects/CacheableTest.java [hunk 9]
```diff
@@ -74,6 +74,32 @@ public void cachesSimpleCall() throws Exception {
...
+            throw new IllegalStateException(ex);
+        }
+        final String thrid = foo.asyncGet().toString();
+        MatcherAssert.assertThat(first, Matchers.equalTo(thrid));
+        try {
+            TimeUnit.SECONDS.sleep(2);
+        } catch (final InterruptedException ex) {
+            throw new IllegalStateException(ex);
+        }
+        final String forth = foo.asyncGet().toString();
```
Comment (pinaf) [inline src/test/java/com/jcabi/aspects/CacheableTest.java]: @54uso we can inline this variable

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java (hunk 1 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: Ok, Thanks.

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 10]
```diff
@@ -265,6 +245,10 @@ private void clean() {
          */
         private final transient MethodCacher.Key key;
         /**
+         * Whether asynchronous update.
+         */
+        private final transient boolean asynchupdate;
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso let's rename this to `async`

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 11]
```diff
@@ -276,19 +260,34 @@ private void clean() {
...
+            this.asynchupdate = aupdate;
         }
+
         @Override
         public String toString() {
             return Mnemos.toText(this.cached, true, false);
         }
+
+        /**
+         * Public ctor.
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso constructor? what is the idea here?

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 12]
```diff
@@ -390,6 +397,15 @@ public boolean expired() {
                 this.level = Loggable.DEBUG;
             }
         }
+
+        /**
+         * Get log level.
+         * @return Log level of current method.
+         */
+        public int getLevel() {
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso isn't protected enough?

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 13]
```diff
@@ -148,13 +136,23 @@ public Object cache(final ProceedingJoinPoint point) throws Throwable {
...
+    /**
      * Flush cache.
      * @param point Join point
      * @return Value of the method
      * @since 0.7.14
      * @deprecated Since 0.7.17, and preflush() should be used
      * @throws Throwable If something goes wrong inside
-     * @checkstyle IllegalThrows (3 lines)
+     * @checkstyle IllegalThrows (4 lines)
+     * @checkstyle MethodsOrderCheck (3 lines)
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso we can get rid of this

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 14]
```diff
@@ -148,13 +136,23 @@ public Object cache(final ProceedingJoinPoint point) throws Throwable {
...
+
+    /**
      * Flush cache.
      * @param point Join point
      * @return Value of the method
      * @since 0.7.14
      * @deprecated Since 0.7.17, and preflush() should be used
      * @throws Throwable If something goes wrong inside
-     * @checkstyle IllegalThrows (3 lines)
+     * @checkstyle IllegalThrows (4 lines)
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso we can revert this

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 15]
```diff
@@ -72,34 +70,19 @@
...
+     * Save the keys which need update.
      */
-    private final transient ScheduledExecutorService cleaner =
-        Executors.newSingleThreadScheduledExecutor(
-            new NamedThreads(
-                "cacheable",
-                "automated cleaning of expired @Cacheable values"
-            )
-        );
+    private final transient BlockingQueue<Key> updatekeys =
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso I see we have this same variable in the new class `UpdateMethodCacher`. Do you think we could get rid of it from here?

**Comment by pinaf** [comment]
@54uso thanks

File: src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java (hunk 6 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java]: Yes. Executing method named "update" Immediately.

**Comment by pinaf** [comment]
@54uso please see 15 comments above

**Comment by pinaf** [comment]
@54uso also, please make sure that every comment you make here is addressed to someone (probably me). thanks again.

File: src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java (hunk 6 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java]: Yes. Executing method named "update" Immediately.

Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java]: @54uso so it is a one-shot action? it doesn't execute again after a time interval?

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java (hunk 11 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: Sorry, my comment is wrong. This method is used to obtain a new instance.

**Comment by 54uso** [comment]
@pinaf Thank u for ur help.  Best wish.

**Comment by pinaf** [comment]
@54uso if you can, don't squash your commit so I can track the changes more easily.

**Comment by 54uso** [comment]
@pinaf ok

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java (hunk 12 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @pinaf checkstyle hinted that final class should not contain protected methods when I used "protected".

File: src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java (hunk 6 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/UpdateMethodCacher.java]: @pinaf Specific action will be execute in the while loop

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java (hunk 15 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: This variable was used in this class, and if we want to get rid of it from here, we must add a new method to the class 'UpdateMethodCacher'.

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java (hunk 2 above)
Comment (54uso) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @pinaf We can get Tunnel by Key, so just to save key

**Comment by 54uso** [comment]
@pinaf I have already modified the code as you suggested. Thanks for your help. Best wishes.

File: src/main/java/com/jcabi/aspects/aj/MethodCacher.java [hunk 16]
```diff
@@ -75,31 +73,16 @@
...
-                new Runnable() {
-                    @Override
-                    public void run() {
-                        MethodCacher.this.clean();
-                    }
-                }
-            ),
-            1L, 1L, TimeUnit.SECONDS
-        );
+        new UpdateMethodCacher(this.tunnels, this.updatekeys).start();
```
Comment (pinaf) [inline src/main/java/com/jcabi/aspects/aj/MethodCacher.java]: @54uso can't the user just instantiate `UpdateMethodCacher` himself if he wants the new functionality? Otherwise he created a plain `MethodCacher`.

**Comment by pinaf** [comment]
@54uso The new functionality should be optional - if the user wants the async behavior, he can instantiate the new class `UpdateMethodCacher` himself. If he doesn't, he should be able to instantiate the old class `MethodCacher` and have the old behavior.

File: src/main/java/com/jcabi/aspects/Cacheable.java [hunk 17]
```diff
@@ -104,6 +104,12 @@
     boolean forever() default false;
 
     /**
+     * Returns the current store after the expiration, and
+     * then asynchronously update the data.
+     */
+    boolean asyncUpdate() default false;
```
Comment (54uso) [inline src/main/java/com/jcabi/aspects/Cacheable.java]: @pinaf The user can decide whether to use  asynchronous method through this property. If asyncUpdate equal to true, using the asynchronous mode, or do not use.

**Comment by pinaf** [comment]
@54uso sure but the class `MethodCacher` is creating an instance of `UpdateMethodCacher`. My question is: do you think we can separate the new functionality completely into the new class?

**Comment by 54uso** [comment]
@pinaf let me think about it. Thanks.

**Comment by 54uso** [comment]
@pinaf  I've no better idea, can you give me some suggestions? Thanks.

**Comment by pinaf** [comment]
@yegor256 Do you think the current design is good enough? The fact that the classes are mixed up doesn't please me but I don't have an alternative solution.

**Comment by yegor256** [comment]
@rultor looks good to me, try to merge

**Comment by dmarkov** [comment]
@pinaf Much appreciated! **25 mins** was added to your account, payment ID `AP-4MS23604KG499291C`, time spent is 498 hours and 24 mins. extra minutes for review comments (c=10). +25 added to your rating, current score is: [+3881](http://www.netbout.com/b/35128?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy
